File: 91-decode-ways/91-decode-ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        @lru_cache
        def dfs(cur):
            if cur == len(s):
                return 1
            if s[cur] == '0':
                return 0
            
            # other case
            res = dfs(cur + 1)
            if cur + 1 < len(s) and int(s[cur: cur + 2]) in range(10, 27):
                res += dfs(cur + 2)
            return res
        
        return dfs(0)

File: 91-decode-ways/91-decode-ways.java
class Solution {
    public int numDecodings(String s) {
        if (s == null || s.length() == 0) {
			return 0;
		}
		char[] str = s.toCharArray();
		int N = str.length;
		int[] dp = new int[N + 1];
		dp[N] = 1;
		for (int i = N - 1; i >= 0; i--) {
			if (str[i] != '0') {
				int ways = dp[i + 1];
				if (i + 1 < str.length && (str[i] - '0') * 10 + str[i + 1] - '0' < 27) {
					ways += dp[i + 2];
				}
				dp[i] = ways;
			}
		}
		return dp[0];
    }
}